.github/scripts/promote_branch.sh

```diff
@@ -0,0 +1,157 @@
+    echo "                      even if the change has not been in staging for one week."
+    echo "  --debug:            For debugging purposes, do everything except for the git"
+    echo "                      push and deleting the temp repo."
+    echo
+    echo "  --target-branch has to be specified."
+}
+
+check_if_branch_differs() {
+    ACTUAL_DIFFERENT_LINES=$(git diff --numstat origin/$1 | wc -l)
+    if [ $ACTUAL_DIFFERENT_LINES -ne 0 ] ; then
+        echo "Lines differ in branch $1"
+        echo "Actual differing lines: $(git diff --numstat origin/$1)"
+        exit 1
+    fi
+}
+
+check_if_any_commits_in_last_week() {
+    NEW_COMMITS=$(git log --oneline --since="$(date --date="7 days ago" +%Y-%m-%d)" | wc -l)
+    if [ $NEW_COMMITS -ne 0 ] ; then
+        echo "There are commits in staging that are less than a week old. Blocking promotion to production"
+        echo "Commits less than a week old: $(git log --oneline --since="$(date --date="7 days ago" +%Y-%m-%d)")"
+        exit 1
+    fi
+}
+
+if [ -z "${TARGET_BRANCH}" ]; then
+    echo -e "Error: missing 'target-branch' argument\n"
+    print_help
+    exit 1
+fi
+if [ "${TARGET_BRANCH}" == "staging" ]; then
+    SOURCE_BRANCH="development"
+elif [ "${TARGET_BRANCH}" == "production" ]; then
+    SOURCE_BRANCH="staging"
+else
+    echo "Invalid target-branch. Only 'staging' and 'production' are allowed"
+    print_help
+    exit 1
+fi
+if [ -z "${GITHUB_TOKEN}" ]; then
+    echo -e "Error: missing 'GITHUB_TOKEN' environment variable\n"
+    print_help
+    exit 1
+fi
+
+# Personal access token with appropriate permissions
+token="${GITHUB_TOKEN}"
+
+# Clone the repository
+tmpDir=$(mktemp -d)
+releaseServiceCatalogDir=${tmpDir}/release-service-catalog
+mkdir -p ${releaseServiceCatalogDir}
+
+echo -e "---\nPromoting release-service-catalog ${SOURCE_BRANCH} to ${TARGET_BRANCH}\n---\n"
+
+git clone "git@github.com:$ORG/$REPO.git" ${releaseServiceCatalogDir}
+cd ${releaseServiceCatalogDir}
+
+# A change cannot go into production if the changes in staging are less than a week old
+if [[ "${TARGET_BRANCH}" == "production" && "${OVERRIDE}" != "true" ]] ; then
+    git checkout origin/staging
+    check_if_any_commits_in_last_week
+fi
+
+# A change cannot go into staging if staging and production differ
+if [[ "${TARGET_BRANCH}" == "staging" && "${FORCE_TO_STAGING}" != "true" ]] ; then
+    git checkout origin/staging
+    update_taskGitRevision_in_pipelines production
+    check_if_branch_differs production
+fi
+
+echo "Included PRs:"
+COMMITS=($(git rev-list --first-parent --ancestry-path origin/"$TARGET_BRANCH"'...'origin/"$SOURCE_BRANCH"))
+## now loop through the above array
+for COMMIT in "${COMMITS[@]}"
+do
+  echo $(curl -s   -H 'Authorization: token  '"$token"  'https://api.github.com/search/issues?q=sha:'"$COMMIT" | jq -r '.items[]
+    | select(.repository_url=="https://api.github.com/repos/'"$ORG"'/'"$REPO"'")
+    | .pull_request | select(.merged_at!=null) | .html_url')
```

Comment:
I was wondering if this could also print the PR titles. So I checked for myself and there is nothing like that in the response. You get for example this:
```sh
curl -s   -H 'Authorization: token  '"$GITHUB_TOKEN" 'https://api.github.com/search/issues?q=sha:4e6e0dc38d57a4fbbc7c21cb66c12446258b409e' | jq -r '.items[] | select(.repository_url=="https://api.github.com/repos/redhat-appstudio/release-service-catalog") | .pull_request'
{
  "url": "https://api.github.com/repos/redhat-appstudio/release-service-catalog/pulls/337",
  "html_url": "https://github.com/redhat-appstudio/release-service-catalog/pull/337",
  "diff_url": "https://github.com/redhat-appstudio/release-service-catalog/pull/337.diff",
  "patch_url": "https://github.com/redhat-appstudio/release-service-catalog/pull/337.patch",
  "merged_at": "2024-01-15T16:33:39Z"
}
```

But this makes me think. We could just print the title for each of the commits, like:
```sh
git show --oneline --no-patch $COMMIT
```
This could either be a separate `for` cycle, or printed as part of the existing loop.

I think that would make it easier to see what was actually included when checking the log.

Reply:
Yes, I like that for sure. I made it part of the same for loop. I had thought of wanting to show more than just the PR title when we went to push instead of PR for promotion, but then totally forgot. Updated

Reply:
OK, cool. I guess we can tweak it further if we feel we need more in the future.